codex/stores/repostore/operations.nim

```diff
@@ -0,0 +1,213 @@
+## Nim-Codex
+## Copyright (c) 2024 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+import pkg/chronos
+import pkg/chronos/futures
+import pkg/datastore
+import pkg/datastore/typedds
+import pkg/libp2p/cid
+import pkg/metrics
+import pkg/questionable
+import pkg/questionable/results
+
+import ./coders
+import ./types
+import ../blockstore
+import ../keyutils
+import ../../blocktype
+import ../../clock
+import ../../logutils
+import ../../merkletree
+
+logScope:
+  topics = "codex repostore"
+
+declareGauge(codex_repostore_blocks, "codex repostore blocks")
+declareGauge(codex_repostore_bytes_used, "codex repostore bytes used")
+declareGauge(codex_repostore_bytes_reserved, "codex repostore bytes reserved")
+
+proc putLeafMetadata*(self: RepoStore, treeCid: Cid, index: Natural, blkCid: Cid, proof: CodexProof): Future[?!StoreResultKind] {.async.} =
+  without key =? createBlockCidAndProofMetadataKey(treeCid, index), err:
+    return failure(err)
+
+  await self.metaDs.modifyGet(key,
+    proc (maybeCurrMd: ?LeafMetadata): Future[(?LeafMetadata, StoreResultKind)] {.async.} =
+      var
+        md: LeafMetadata
+        res: StoreResultKind
+
+      if currMd =? maybeCurrMd:
+        md = currMd
+        res = AlreadyInStore
+      else:
+        md = LeafMetadata(blkCid: blkCid, proof: proof)
+        res = Stored
+
+      (md.some, res)
+  )
+
+proc getLeafMetadata*(self: RepoStore, treeCid: Cid, index: Natural): Future[?!LeafMetadata] {.async.} =
+  without key =? createBlockCidAndProofMetadataKey(treeCid, index), err:
+    return failure(err)
+
+  without leafMd =? await get[LeafMetadata](self.metaDs, key), err:
+    if err of DatastoreKeyNotFound:
+      return failure(newException(BlockNotFoundError, err.msg))
+    else:
+      return failure(err)
+
+  success(leafMd)
+
+proc updateTotalBlocksCount*(self: RepoStore, plusCount: Natural = 0, minusCount: Natural = 0): Future[?!void] {.async.} =
+  await self.metaDs.modify(CodexTotalBlocksKey,
+    proc (maybeCurrCount: ?Natural): Future[?Natural] {.async.} =
+      let count: Natural =
+        if currCount =? maybeCurrCount:
+          currCount + plusCount - minusCount
+        else:
+          plusCount - minusCount
+
+      self.totalBlocks = count
+      codex_repostore_blocks.set(count.int64)
+      count.some
+  )
+
+proc updateQuotaUsage*(
```

Comment:
Same as above, probably makes sense to return the new counts?

Reply:
I would have the same argument as above, so let's just keep the discussion in the thread above.